tests/prague/eip6206_jumpf/test_jumpf_stack.py

```diff
@@ -0,0 +1,130 @@
+"""
+EOF JUMPF tests covering stack validation rules.
+"""
+import pytest
+
+from ethereum_test_tools import EOFException, EOFTestFiller, StateTestFiller
+from ethereum_test_tools.eof.v1 import Container, Section
+from ethereum_test_tools.eof.v1.constants import NON_RETURNING_SECTION
+from ethereum_test_tools.vm.opcode import Opcodes as Op
+
+from .helpers import execute_tests
+from .spec import EOF_FORK_NAME
+
+REFERENCE_SPEC_GIT_PATH = "EIPS/eip-6206.md"
+REFERENCE_SPEC_VERSION = "a1775816657df4093787fb9fe83c2f7cc17ecf47"
+
+pytestmark = pytest.mark.valid_from(EOF_FORK_NAME)
+
+
+@pytest.mark.parametrize(
+    "target_inputs",
+    [0, 2, 4],
+    ids=lambda x: "ti-%d" % x,
+)
+@pytest.mark.parametrize(
+    "stack_height",
+    [0, 2, 4],
+    ids=lambda x: "h-%d" % x,
+)
+def test_jumpf_stack_non_returning_rules(
+    state_test: StateTestFiller,
+    eof_test: EOFTestFiller,
+    target_inputs: int,
+    stack_height: int,
+):
+    """
+    Tests for JUMPF validation stack rules.  Non-returning section cases.
+    Valid cases are executed.
+    """
+    container = Container(
+        name="stack-non-retuning_h-%d_ti-%d" % (stack_height, target_inputs),
+        sections=[
+            Section.Code(
+                code=Op.JUMPF[1],
+                code_outputs=NON_RETURNING_SECTION,
+            ),
+            Section.Code(
+                code=Op.PUSH0 * stack_height + Op.JUMPF[2],
+                code_outputs=NON_RETURNING_SECTION,
+                max_stack_height=stack_height,
+            ),
+            Section.Code(
+                code=Op.POP * target_inputs + Op.SSTORE(0, 1) + Op.STOP,
+                code_inputs=target_inputs,
+                code_outputs=NON_RETURNING_SECTION,
+                max_stack_height=max(2, target_inputs),
+            ),
+        ],
+    )
+
+    if stack_height < target_inputs:
+        container.validity_error = EOFException.UNDEFINED_EXCEPTION
+
+    execute_tests(state_test, eof_test, container)
+
+
+@pytest.mark.parametrize(
+    "current_outputs",
+    [0, 2, 4],
+    ids=lambda x: "co-%d" % x,
+)
+@pytest.mark.parametrize(
+    "target_outputs",
+    [0, 2, 4],
+    ids=lambda x: "to-%d" % x,
+)
+@pytest.mark.parametrize(
+    "target_inputs",
+    [0, 2, 4],
+    ids=lambda x: "to-%d" % x,
+)
+@pytest.mark.parametrize("stack_diff", [-1, 0, 1], ids=["less-stack", "same-stack", "more-stack"])
+def test_jumpf_stack_returning_rules(
+    state_test: StateTestFiller,
+    eof_test: EOFTestFiller,
+    current_outputs: int,
+    target_outputs: int,
+    target_inputs: int,
+    stack_diff: int,
+):
+    """
+    Tests for JUMPF validation stack rules.  Returning section cases.
+    Valid cases are executed.
+    """
+    if target_outputs > current_outputs:
+        # These create invalid containers without JUMPF validation, Don't test.
+        return
+    if target_inputs == 0 and stack_diff < 0:
```

Comment:
you can combine parameters into vectors. 

```
@pytest.mark.parametrize(
    "current_output, target_output",
    [(0, 0), (2, 2), (4, 4)],
    ids=lambda c, t: f"co-{c}_to-{t}"
)
```
so you can exclude unintended intersections of the vectors from testing

Reply:
That would be 18 pairs (9 by two sets) before removing collisions.  The advantage I see of keeping them separate and rejecting specific combinations is that in the test we can explain why the combo is invalid, rather than the combo magically not existing.